alibi_detect/cd/keops/mmd.py

```diff
@@ -0,0 +1,179 @@
+            _ = self.kernel(LazyTensor(x[:, None, :]), LazyTensor(x[None, :, :]), infer_sigma=self.infer_sigma)
+            self.infer_sigma = False
+        else:
+            self.infer_sigma = True
+
+    def _mmd2(self, x_all: torch.Tensor, perms: List[torch.Tensor], m: int, n: int) \
+            -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Batched (across the permutations) MMD^2 computation for the original test statistic and the permutations.
+
+        Parameters
+        ----------
+        x_all
+            Concatenated reference and test instances.
+        perms
+            List with permutation vectors.
+        m
+            Number of reference instances.
+        n
+            Number of test instances.
+
+        Returns
+        -------
+        MMD^2 statistic for the original and permuted reference and test sets.
+        """
+        k_xx, k_yy, k_xy = [], [], []
+        for batch in range(self.n_batches):
+            i, j = batch * self.batch_size, (batch + 1) * self.batch_size
+            # construct stacked tensors with a batch of permutations for the reference set x and test set y
+            x = torch.cat([x_all[perm[:m]][None, :, :] for perm in perms[i:j]], 0)
+            y = torch.cat([x_all[perm[m:]][None, :, :] for perm in perms[i:j]], 0)
+            if batch == 0:
+                x = torch.cat([x_all[None, :m, :], x], 0)
+                y = torch.cat([x_all[None, m:, :], y], 0)
+            x, y = x.to(self.device), y.to(self.device)
+
+            # batch-wise kernel matrix computation over the permutations
+            k_xx.append(self.kernel(
+                LazyTensor(x[:, :, None, :]), LazyTensor(x[:, None, :, :])).sum(1).sum(1).squeeze(-1))
+            k_yy.append(self.kernel(
+                LazyTensor(y[:, :, None, :]), LazyTensor(y[:, None, :, :])).sum(1).sum(1).squeeze(-1))
+            k_xy.append(self.kernel(
+                LazyTensor(x[:, :, None, :]), LazyTensor(y[:, None, :, :])).sum(1).sum(1).squeeze(-1))
+        c_xx, c_yy, c_xy = 1 / (m * (m - 1)), 1 / (n * (n - 1)), 2. / (m * n)
+        stats = c_xx * (torch.cat(k_xx) - m) + c_yy * (torch.cat(k_yy) - n) - c_xy * torch.cat(k_xy)
+        return stats[0], stats[1:]
+
+    def score(self, x: Union[np.ndarray, list]) -> Tuple[float, float, float]:
+        """
+        Compute the p-value resulting from a permutation test using the maximum mean discrepancy
+        as a distance measure between the reference data and the data to be tested.
+
+        Parameters
+        ----------
+        x
+            Batch of instances.
+
+        Returns
+        -------
+        p-value obtained from the permutation test, the MMD^2 between the reference and test set,
+        and the MMD^2 threshold above which drift is flagged.
+        """
+        x_ref, x = self.preprocess(x)
+        x_ref = torch.from_numpy(x_ref).float()  # type: ignore[assignment]
+        x = torch.from_numpy(x).float()  # type: ignore[assignment]
```

Comment:
I note that in the `torch` implementation this are also sent to `self.device`

Reply:
Yes here we don't want to do that as we're stacking all the permutations together and then optionally operate on slices of permutations which are placed on the device. Check the `_mmd2` function for the logic.